# Task 1- Open Images class hierarchy 

Here I have used the given annotation files and the csv file from the given data.Reading the csv dataset using pandas.

In [1]:
import pandas as pd
cls=pd.read_csv('oidv6-class-descriptions.csv', header=None)
classes2name={i:j for i,j in zip(cls[0], cls[1])}
name2classes={j:i for i,j in zip(cls[0], cls[1])}
cls.tail()

0                    1
19990  /m/0zrpfhj             Throwing
19991  /m/0zrrls2           Red dahlia
19992  /m/0zrthkd         Brine (Food)
19993  /m/0zrv5th  Lamborghini huracán
19994    /m/0zvk5               Helmet

Reading the JSON file from the given dataset to understand the heirarchy level as per the name indentation.

In [3]:
import json
hier = json.load(open('bbox_labels_600_hierarchy.json','r'))
level1=[]
level2=[]
level3=[]
for l2 in hier['Subcategory']:
    print(classes2name[l2['LabelName']])
    level3.append(classes2name[l2['LabelName']])
    try:
        for j in l2['Subcategory']:
            print('----> ',classes2name[j['LabelName']])
            level2.append(classes2name[j['LabelName']])
            try:
                for k in j['Subcategory']:
                    print('\t----> ',classes2name[k['LabelName']])
                    level1.append(classes2name[k['LabelName']])
            except:
                pass
    except:
        pass
        
level1 = set(level1)
level2 = set(level2)
level3 = set(level3)


Coin
Flag
Light bulb
Toy
---->  Doll
---->  Balloon
---->  Dice
---->  Flying disc
---->  Kite
---->  Teddy bear
Home appliance
---->  Washing machine
---->  Toaster
---->  Oven
---->  Blender
---->  Gas stove
---->  Mechanical fan
---->  Heater
---->  Kettle
---->  Hair dryer
---->  Refrigerator
---->  Wood-burning stove
---->  Humidifier
---->  Mixer
---->  Coffeemaker
---->  Microwave oven
---->  Dishwasher
---->  Sewing machine
---->  Hand dryer
---->  Ceiling fan
Plumbing fixture
---->  Sink
---->  Bidet
---->  Shower
---->  Tap
---->  Bathtub
---->  Toilet
Office supplies
---->  Scissors
---->  Poster
---->  Calculator
---->  Box
---->  Stapler
---->  Whiteboard
---->  Pencil sharpener
---->  Eraser
---->  Fax
---->  Adhesive tape
---->  Ring binder
---->  Pencil case
---->  Plastic bag
---->  Paper cutter
---->  Toilet paper
---->  Envelope
---->  Pen
Paper towel
Pillow
Kitchenware
---->  Kitchen utensil
	---->  Chopsticks
	---->  Ladle
	---->  Spatula
	---->  Can opener
	----> 

---->  Cannon
---->  Dagger
---->  Knife
---->  Rifle
---->  Shotgun
---->  Tank
---->  Axe
---->  Handgun
---->  Sword
---->  Missile
---->  Bomb
Cassette deck
Headphones
Laptop
Computer keyboard
Printer
Computer mouse
Computer monitor
Power plugs and sockets
Light switch
Television
Telephone
---->  Mobile phone
---->  Corded phone
Tablet computer
Microphone
Ipod
Remote control
Drink
---->  Beer
---->  Cocktail
---->  Coffee
---->  Juice
---->  Tea
---->  Wine
Bicycle wheel
Door handle
Door
Window
Stairs
Porch
Human eye
Skull
Human head
Human face
Human mouth
Human ear
Human nose
Human hair
Human hand
Human foot
Human arm
Human leg
Human beard
Human body
Auto part
---->  Vehicle registration plate
---->  Wheel
---->  Seat belt
---->  Tire
Coconut


# Classes count
We can see that we obtain 3 levels of classes hierarchy.

In [4]:
print('Classes count in level 1 is {}'.format(len(level1)))
print('Classes count in level 2 is {}'.format(len(level2)))
print('Classes count in level 3 is {}'.format(len(level3)))
print('Total unique class counts are {}'.format(len(level1)+len(level2)+len(level3)))

Classes count in level 1 is 235
Classes count in level 2 is 252
Classes count in level 3 is 90
Total unique class counts are 577


So there are 577 unique classes that we obtain from the below classes :
ancestor classes : 90
parent class : 235
siblings class :252

To check  if both parent class and sibling class belong to the same ancestor class(es) 

In [5]:
print('level1 and level2 overlaps = {}'.format(len(level2&level1)))
print('level2 and level3 overlaps = {}'.format(len(level2&level3)))
print('level1 and level3 overlaps = {}'.format(len(level3&level1)))

level1 and level2 overlaps = 19
level2 and level3 overlaps = 0
level1 and level3 overlaps = 0


In [15]:
cls[1].value_counts()

Lilac                2
Wave                 2
Friendship day       2
Slate                2
Swim brief           2
                    ..
Grapefruit           1
Polikarpov po-2      1
Bangers and mash     1
Japanese sea bass    1
Nasi kandar          1
Name: 1, Length: 19983, dtype: int64